GraphPattern.pde

```diff
@@ -974,3 +974,237 @@ public class DDTwinkle extends GraphPattern {
+      // recent versions of muse-io look like they don't catch those values any longer :(
+      // calm = muse.getTheta()/muse.getAlpha();
+      // attention = muse.getBeta()/muse.getAlpha();
+      //  // Placeholders, use these for now?
+      calm = muse.getAlpha();
+      attention = muse.getGamma();
+      fadeRate = map(calm, 0.0, 1.0, (float)fade.range.min, (float)fade.range.max);
+      speedRate = map(attention, 0.0, 1.0, (float)speed.range.min, (float)speed.range.max);
+    }
+    else {
+      fadeRate = fade.getValuef();
+      speedRate = speed.getValuef();
+    }
+
+    // Scale brightness for all points
+    for (LXPoint p : model.points) {
+     colors[p.index] =
+         LXColor.scaleBrightness(colors[p.index], fadeRate);
+    }
+
+    // Ensure all living pixies have the latest color
+    for (TtlPixie p : this.pixies) {
+      p.updateColor();
+    }
+
+    // Ensure all solid bars have the latest color.
+    anglemod=anglemod+1;
+    if (anglemod > 360)
+      anglemod = anglemod % 360;
+    for (Bar bar : this.solidBars) {
+      for (LXPoint point: bar.points) {
+        colors[point.index]=lx.hsb(((atan(point.x/point.z))*360/PI+anglemod),80,50);
+      }
+    }
+
+    // To avoid concurrent modification exceptions while iterating
+    // over living pixies, we accumulate pixies to add/remove in these lists.
+    List<TtlPixie> newPixies = new ArrayList<TtlPixie>();
+    List<TtlPixie> removePixies = new ArrayList<TtlPixie>();
+
+    ALL_PIXIES_ITER:
+    for (TtlPixie p : this.pixies) {
+
+      Bar thisBar = model.getBar(p.fromNode, p.toNode);
+
+      // If a pixie's TTL hits 0, remove it.
+      if (p.ttl <= 0) {
+        removePixies.add(p);
+        barPixies.get(thisBar).remove(p);
+        continue;
+      }
+
+      double drawOffset = p.offset;
+      p.offset += (deltaMs / 1000.0) * speedRate;
+
+      // Check the bar in the opposite direction for
+      // pixies that are colliding with this pixie. Note that we
+      // only check if for collisions if a collision *hasn't*
+      // already occurred on this bar.
+      Bar oppositeBar = model.getBar(p.toNode, p.fromNode);
+      if (!solidBars.contains(thisBar)) {
+        for (TtlPixie oncomingPixie : (Set<TtlPixie>)barPixies.get(oppositeBar)) {
+          // Note the shenanery here: offset 0 in one direction is
+          // opposite offset 0 in the other, so one of them needs to be flipped
+          // before we can compare their "physical" locations.
+          if (Math.abs((int)oncomingPixie.offset - (thisBar.points.length - (int)p.offset)) <= 2) {
+            p.isCollided = true;
+            oncomingPixie.isCollided = true;
+
+            solidBars.add(thisBar);
+            solidBars.add(oppositeBar);
+            blankBars.remove(thisBar);
+            blankBars.remove(oppositeBar);
+
+            // When two pixies collide, spawn a new one.
+            addNewPixieToSet(newPixies);
+
+            continue ALL_PIXIES_ITER;
```

Comment:
Hmm. This GOTO gives me that code smell feel. Gonna let it live, but better would be to avoid using control structures like this. It tends to obfuscate the intended code flow, and hides potential bugs.

Reply:
1) Feature branches can only be created by individuals with write access to the repo, which I do not have.
2) Labeled blocks are quite common in Java; using something else i.e. a method call would simply waste stack space. For examples of how common labeled blocks are, see line 1900 of JDK's `String` class: http://www.docjar.com/html/api/java/lang/String.java.html

Reply:
You now have write access! :)

Thanks for the info on the labeled block, handy to know!